rootfs/etc/nginx/lua/statsd.lua

```diff
@@ -0,0 +1,174 @@
+    ngx.ctx.statsd_sockets = ctx
+  end
+
+  local sock = ctx[id]
+  if not sock then
+    sock, err = create_udp_socket(host, port)
+    if not sock then
+      return nil, err
+    end
+
+    ctx[id] = sock
+  end
+
+  return sock, nil
+end
+
+local function generate_tag_string(tags)
+  if not tags or util_tablelength(tags) == 0 then
+    return ""
+  end
+
+  local tag_str = default_tag_string
+  for k,v in pairs(tags) do
+    if string_len(tag_str) > 2 then
+      tag_str = tag_str .. ","
+    end
+    tag_str = tag_str .. k .. ":" .. v
+  end
+
+  return tag_str
+end
+
+local function generate_packet(metric, key, value, tags, sampling_rate)
+  if sampling_rate == 1 then
+    sampling_rate = ""
+  else
+    sampling_rate = string_format("|@%g", sampling_rate)
+  end
+
+  return string_format("%s:%s|%s%s%s", key, tostring(value), metric, sampling_rate, generate_tag_string(tags))
+end
+
+local function metric(metric_type, key, value, tags, sample_rate)
+  if not value then
+    return nil, "no value passed"
+  end
+  if value == '-' then
+    return nil, nil -- don't pass an error to avoid logging to error log
+  end
+
+  if not _M.config then
+    return true, nil
+  end
+
+  local sampling_rate = sample_rate or _M.config.sampling_rate
+
+  if sampling_rate ~= 1 and math.random() > sampling_rate then
+    return nil, nil -- don't pass an error to avoid logging to error log
+  end
+
+  local packet = generate_packet(metric_type, key, value, tags, sampling_rate)
+
+  local sock, err = get_udp_socket(_M.config.host, _M.config.port)
+  if not sock then
+    return nil, err
+  end
+
+  return sock:send(packet)
+end
+
+local function send_metrics(...)
+  local ok, err = metric(...)
+  if not ok and err then
+    ngx.log(ngx.WARN, "failed logging to statsd: " .. tostring(err))
+  end
+  return ok, err
+end
+
+-- to avoid logging everywhere in #metric
+local function log_metric(...)
+  local ok, err = defer.to_timer_phase(send_metrics, ...)
+  if not ok then
+    local msg = "failed to log metric: " .. tostring(err)
+    ngx.log(ngx.ERR,  msg)
+    return nil, msg
+  end
+  return true
+end
+
+-- Statsd module level convenince functions
+
+function _M.increment(key, value, tags, ...)
+  return log_metric(METRIC_COUNTER, key, value or 1, tags, ...)
+end
+
+function _M.gauge(key, value, tags, ...)
+  return log_metric(METRIC_GAUGE, key, value, tags, ...)
+end
+
+function _M.histogram(key, value, tags, ...)
+  return log_metric(METRIC_HISTOGRAM, key, value, tags, ...)
+end
+
+function _M.set(key, value, tags, ...)
+  return log_metric(METRIC_SET, key, value, tags, ...)
+end
+
+function _M.time(f)
+  local start_time = ngx.now()
+  local ret = { f() }
+  return ret, (ngx.now() - start_time) * MICROSECONDS
+end
+
+function _M.measure(key, f, tags)
+  local ret, time = _M.time(f)
+  _M.histogram(key, time, tags or {})
+  return unpack(ret)
+end
+
+_M.config = {
+  host = os.getenv("STATSD_HOST") or '127.0.0.1',
+  port = os.getenv("STATSD_PORT") or '8125',
```

Comment:
Do we want to fail here instead of resorting to defaults here @ElvinEfendi?

Reply:
Could either log a warning if either env var are empty and proceed - or we could error out. WDYT?

Reply:
@andrewlouis93 I think we should `error`